File: lib/services/AuthService.dart
// lib/services/AuthService.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

import 'package:google_sign_in/google_sign_in.dart';
import 'package:flutter_facebook_auth/flutter_facebook_auth.dart';
import 'package:kakao_flutter_sdk_user/kakao_flutter_sdk_user.dart' as kakao;
import 'package:flutter_naver_login/flutter_naver_login.dart';

class AuthService {
  final FirebaseAuth _auth;
  final FirebaseFirestore _fs;
  final FlutterSecureStorage _storage;

  AuthService({
    FirebaseAuth? auth,
    FirebaseFirestore? fs,
    FlutterSecureStorage? storage,
  })  : _auth = auth ?? FirebaseAuth.instance,
        _fs = fs ?? FirebaseFirestore.instance,
        _storage = storage ?? const FlutterSecureStorage();

  // 상태
  User? get currentUser => _auth.currentUser;
  String? get currentUserId => _auth.currentUser?.uid;
  Stream<User?> get authStateChanges => _auth.authStateChanges();
  Future<bool> isLoggedIn() async => currentUser != null;

  // 토큰
  Future<String?> getIdToken({bool forceRefresh = false}) async {
    final u = _auth.currentUser;
    if (u == null) return null;
    return await u.getIdToken(forceRefresh);
  }

  Future<void> saveIdToken() async {
    final t = await getIdToken(forceRefresh: true);
    if (t != null) await _storage.write(key: 'auth_token', value: t);
  }

  Future<bool> verifyStoredIdToken() async {
    final stored = await _storage.read(key: 'auth_token');
    if (stored == null || currentUser == null) return false;
    try {
      final now = await getIdToken();
      return stored == now;
    } catch (e) {
      debugPrint('verifyStoredIdToken error: $e');
      return false;
    }
  }

  // Firestore 사용자 문서 생성(최초 로그인 1회)
  Future<void> registerUser({
    required String userName,
    required String countryCode,
    required String countryName,
    String? profileImageUrl,
    String? bio,
  }) async {
    final uid = currentUserId;
    if (uid == null) return;

    final ref = _fs.collection('users').doc(uid);
    final snap = await ref.get();
    if (snap.exists) return;

    final providerId = currentUser?.providerData.isNotEmpty == true
        ? currentUser!.providerData.first.providerId
        : 'password';

    await ref.set({
      'user_id': uid,
      'email': currentUser?.email ?? '',
      'user_name': userName,
      'profile_image': profileImageUrl ?? '',
      'bio': bio ?? '',
      'country_code': countryCode,
      'country_name': countryName,
      'provider': providerId,
      'created_at': FieldValue.serverTimestamp(),
    });
  }

  // 로그아웃
  Future<void> signOut() async {
    try { await GoogleSignIn().signOut(); } catch (_) {}
    try { await FacebookAuth.instance.logOut(); } catch (_) {}
    try { await FlutterNaverLogin.logOut(); } catch (_) {}
    try { await kakao.UserApi.instance.logout(); } catch (_) {}
    await _auth.signOut();
    await _storage.delete(key: 'auth_token');
  }

  // ── Providers ─────────────────────────────────

  Future<UserCredential?> signInWithGoogle() async {
    try {
      final gUser = await GoogleSignIn().signIn();
      if (gUser == null) return null;
      final gAuth = await gUser.authentication;
      final cred = GoogleAuthProvider.credential(
        accessToken: gAuth.accessToken,
        idToken: gAuth.idToken,
      );
      return await _auth.signInWithCredential(cred);
    } catch (e) {
      debugPrint('signInWithGoogle error: $e');
      return null;
    }
  }

  Future<UserCredential?> signInWithFacebook() async {
    try {
      final res = await FacebookAuth.instance.login();
      if (res.status != LoginStatus.success || res.accessToken == null) return null;
      final cred = FacebookAuthProvider.credential(res.accessToken!.token);
      return await _auth.signInWithCredential(cred);
    } catch (e) {
      debugPrint('signInWithFacebook error: $e');
      return null;
    }
  }

  // Kakao/Naver는 백엔드에서 Firebase Custom Token 발급 필요
  Future<UserCredential?> signInWithKakao({
    Future<String> Function(String kakaoAccessToken)? fetchFirebaseCustomToken,
  }) async {
    try {
      final token = await (await kakao.isKakaoTalkInstalled()
          ? kakao.UserApi.instance.loginWithKakaoTalk()
          : kakao.UserApi.instance.loginWithKakaoAccount());

      if (fetchFirebaseCustomToken == null) {
        debugPrint('Kakao accessToken=${token.accessToken}');
        return null;
      }
      final customToken = await fetchFirebaseCustomToken(token.accessToken);
      return await _auth.signInWithCustomToken(customToken);
    } catch (e) {
      debugPrint('signInWithKakao error: $e');
      return null;
    }
  }

  Future<UserCredential?> signInWithNaver({
    Future<String> Function(String naverAccessToken)? fetchFirebaseCustomToken,
  }) async {
    try {
      final res = await FlutterNaverLogin.logIn();
      if (res.status != NaverLoginStatus.loggedIn) return null;

      if (fetchFirebaseCustomToken == null) {
        debugPrint('Naver accessToken=${res.accessToken.accessToken}');
        return null;
      }
      final customToken = await fetchFirebaseCustomToken(res.accessToken.accessToken);
      return await _auth.signInWithCustomToken(customToken);
    } catch (e) {
      debugPrint('signInWithNaver error: $e');
      return null;
    }
  }
  
  Future<UserCredential?> signInWithEmailPassword({
    required String email,
    required String password,
  }) async {
    try {
      return await _auth.signInWithEmailAndPassword(
        email: email,
        password: password,
      );
    } on FirebaseAuthException catch (e) {
      debugPrint('signInWithEmailPassword error: $e');
      rethrow;
    }
  }

  Future<bool> sendEmailVerification() async {
    final u = _auth.currentUser;
    if (u == null) return false;
    await u.reload();
    if (u.emailVerified) return true; // 이미 인증됨
    await u.sendEmailVerification();
    return true;
  }
}

File: ting-rn/src/services/authService.ts
import auth from '@react-native-firebase/auth';
import firestore from '@react-native-firebase/firestore';
import * as SecureStore from 'expo-secure-store';
import { GoogleSignin } from '@react-native-google-signin/google-signin';

const TOKEN_KEY = 'auth_token';
const HAS_LOGGED_IN_KEY = 'has_logged_in_before';

export const authService = {
  get currentUser() {
    return auth().currentUser;
  },

  get currentUserId() {
    return auth().currentUser?.uid ?? null;
  },

  onAuthStateChanged(cb: (user: ReturnType<typeof auth>['currentUser']) => void) {
    return auth().onAuthStateChanged(cb);
  },

  async getIdToken(forceRefresh = false): Promise<string | null> {
    const u = auth().currentUser;
    if (!u) return null;
    return u.getIdToken(forceRefresh);
  },

  async saveIdToken(): Promise<void> {
    const token = await this.getIdToken(true);
    if (token) {
      await SecureStore.setItemAsync(TOKEN_KEY, token);
    }
  },

  async verifyStoredIdToken(): Promise<boolean> {
    const stored = await SecureStore.getItemAsync(TOKEN_KEY);
    if (!stored || !auth().currentUser) return false;
    try {
      const now = await this.getIdToken();
      return stored === now;
    } catch {
      return false;
    }
  },

  async registerUser(params: {
    userName: string;
    countryCode: string;
    countryName: string;
    profileImageUrl?: string;
    bio?: string;
  }): Promise<void> {
    const uid = this.currentUserId;
    if (!uid) return;

    const ref = firestore().collection('users').doc(uid);
    const snap = await ref.get();
    if (snap.exists()) return;

    const providerData = auth().currentUser?.providerData;
    const providerId =
      providerData && providerData.length > 0
        ? providerData[0].providerId
        : 'password';

    await ref.set({
      user_id: uid,
      email: auth().currentUser?.email ?? '',
      user_name: params.userName,
      profile_image: params.profileImageUrl ?? '',
      bio: params.bio ?? '',
      country_code: params.countryCode,
      country_name: params.countryName,
      provider: providerId,
      created_at: firestore.FieldValue.serverTimestamp(),
    });
  },

  async signOut(): Promise<void> {
    try {
      await GoogleSignin.signOut();
    } catch {
      // ignore if Google sign-out fails
    }
    await auth().signOut();
    await SecureStore.deleteItemAsync(TOKEN_KEY);
  },

  // --- Social providers (stubs — wire up in Phase 1) ---

  async signInWithGoogle(): Promise<boolean> {
    try {
      GoogleSignin.configure({
        webClientId:
          '1088275016090-h5uuerbcan6kmskqe0rudflke5c5a95h.apps.googleusercontent.com',
        iosClientId:
          '1088275016090-re9fu4ssqd2iti97kfmqjgti3k0hm2qm.apps.googleusercontent.com',
      });
      await GoogleSignin.hasPlayServices();
      const response = await GoogleSignin.signIn();
      const idToken = response.data?.idToken;
      if (!idToken) return false;

      const credential = auth.GoogleAuthProvider.credential(idToken);
      await auth().signInWithCredential(credential);
      await this.saveIdToken();
      await this.markHasLoggedInBefore();
      return true;
    } catch (e) {
      console.error('Google sign-in error:', e);
      return false;
    }
  },

  async signInWithFacebook(): Promise<boolean> {
    // TODO: implement with react-native-fbsdk-next
    return false;
  },

  async hasLoggedInBefore(): Promise<boolean> {
    return (await SecureStore.getItemAsync(HAS_LOGGED_IN_KEY)) === 'true';
  },

  async markHasLoggedInBefore(): Promise<void> {
    await SecureStore.setItemAsync(HAS_LOGGED_IN_KEY, 'true');
  },
};
